### PR TITLE
add Polymer.dom calls rule

### LINKED ISSUE #143: Polymer.dom warnings
Could easily warn about usages of `Polymer.dom` and potentially fix usages.

**Examples:**
- `Polymer.dom(event).localTarget` to `event.target`
- `Polymer.dom(event).path` to `event.composedPath()`
- `Polymer.dom(event).rootTarget` to `event.composedPath()[0]`

### PR DESCRIPTION
**Pull request by 43081j**
Fixes #143.

---

I was having a look around the linter and this one seemed simple enough to implement. 

Couple of questions:
* Is `Polymer.dom` actually deprecated? if so, should i update the messages to say so?
* I put this in the polymer 2 ruleset as 1.x and hybrid elements still depend on it, right?
* If we want to implement fixes/replacements, we need some way of turning an expression into a string and i don't think the analyser exposes our util method for that (e.g. convert the ast of `some.variable.here` to the literal string `"some.variable.here"`, and more complex cases too)

---
This change is [Reviewable](https://reviewable.io/reviews/polymer/polymer-linter/154)

### CODE REVIEW

File: src/polymer/dom-calls-to-native.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      babel.isIdentifier(expr.object.callee.object) &&
+      babel.isIdentifier(expr.object.callee.property) &&
+      expr.object.callee.object.name === 'Polymer' &&
+      expr.object.callee.property.name === 'dom';
+  }
+}
+
+registry.register(new DomCallsToNative());
+
+function getParsedDocumentContaining(
```
Comment (43081j) [inline src/polymer/dom-calls-to-native.ts]: this thing needs to go away... we have a method now in the analyzer (somewhere inside `Analysis`). but im not sure how we get at it...

**Comment by stramel** [comment]
@43081j my hero 👍🏼

File: src/polymer/dom-calls-to-native.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
```
Comment (stramel) [inline src/polymer/dom-calls-to-native.ts]: Technically I think this should be the year this file was created.

File: src/test/polymer/dom-calls-to-native.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
```
Comment (stramel) [inline src/test/polymer/dom-calls-to-native.ts]: Same here

File: src/polymer/dom-calls-to-native.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+import babelTraverse from 'babel-traverse';
+import * as babel from 'babel-types';
+import {Document, Severity, Warning, SourceRange, ParsedDocument, isPositionInsideRange} from 'polymer-analyzer';
+
+import {registry} from '../registry';
+import {Rule} from '../rule';
+import {stripWhitespace} from '../util';
+
+import {stripIndentation} from '../util';
```
Comment (stramel) [inline src/polymer/dom-calls-to-native.ts]: Can you consolidate these imports?

**Comment by 43081j** [comment]
@stramel i've gone ahead and sorted those things

& also ran the formatter on it to clean up some whitespace etc.

File: src/polymer/dom-calls-to-native.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+              getParsedDocumentContaining(doc.sourceRange, document);
+
+          if (!containingDoc) {
+            return;
+          }
+
+          const sourceRange = containingDoc.sourceRangeForNode(path.node)!;
+
+          if (this._isPolymerDomCall(path.node) &&
+              this._replacements.has(name)) {
```
Comment (rictic) [inline src/polymer/dom-calls-to-native.ts]: Let's move these checks up above the work to get the containing doc, as they're very fast checks, and they don't allocate, whereas containingDoc and sourceRangeForNode are a bit more expensive.

Early exiting with `return` if these don't match is nice too.

File: src/polymer/dom-calls-to-native.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          const containingDoc =
+              getParsedDocumentContaining(doc.sourceRange, document);
+
+          if (!containingDoc) {
+            return;
+          }
+
+          const sourceRange = containingDoc.sourceRangeForNode(path.node)!;
+
+          if (this._isPolymerDomCall(path.node) &&
```
Comment (rictic) [inline src/polymer/dom-calls-to-native.ts]: I think it'd make a bit more sense if we passed in `path.node.object` to `_isPolymerDomCall`

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
 <!-- ## Unreleased -->
```
Comment (rictic) [inline CHANGELOG.md]: Uncomment out the `## Unreleased` section please

Comment (rictic) [inline CHANGELOG.md]: Uncomment out the `## Unreleased` section please

**Review by rictic — COMMENTED** [review]
Very cool! This will be a huge help for people migrating!

A few small suggestions and this is good to land

**Comment by rictic** [comment]
Ignore the node travis failure, that's node v6 getting TypeScript 2.7 which is a bit more strict on us

**Comment by 43081j** [comment]
@rictic i've updated as per your comments.

do you know if i can get the containing doc function from the analyzer somehow? it feels a bit hacky having to pretty much re-define it here

File: src/polymer/dom-calls-to-native.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+          const containingDoc =
+              getParsedDocumentContaining(doc.sourceRange, document);
+
+          if (!containingDoc) {
+            return;
+          }
+
+          const sourceRange = containingDoc.sourceRangeForNode(path.node)!;
+          const replacement = this._replacements.get(name)!;
```
Comment (rictic) [inline src/polymer/dom-calls-to-native.ts]: The ! is a code smell. Consider:

```typescript
const replacement = this._replacements.get(name);

 if (!this._isPolymerDomCall(path.node.object) || replacement === undefined) {
  return;
}
```

and similarly early exiting if sourceRange is undefined (I know there's no current way for that source range to be undefined, but good type checking has caught a lot of other things that didn't seem likely to change at the time that I prefer to just lean heavily on the type checker wherever we can)

File: src/polymer/dom-calls-to-native.ts (hunk 1 above)
Comment (rictic) [reply src/polymer/dom-calls-to-native.ts]: Yeah, we need to pass down the `analysis` to `checkDocument`. Mind putting a TODO and I'll tackle that in a followup PR?

**Review by rictic — COMMENTED** [review]
Just two more super quick things, if you don't mind!

**Comment by 43081j** [comment]
here you go

**Review by rictic — APPROVED** [review]

**Comment by rictic** [comment]
Awesome!

**Pull request merged**